src/core/builders/create-form-validator.spec.ts

```diff
@@ -847,4 +856,181 @@ describe("createFormValidator", () => {
+      }),
+      validate(Schema.choice, x => (x === "A" ? "INVALID_VALUE" : null)),
+    ]);
+
+    const getValue = (field: FieldDescriptor<any> | string): any => {
+      const path = typeof field === "string" ? field : impl(field).__path;
+      switch (path) {
+        case "arrayObjectString":
+          return [{ str: "sm" }, { str: "" }, { str: "valid string" }];
+        case "arrayObjectString[0]":
+          return { str: "sm" };
+        case "arrayObjectString[1]":
+          return { str: "" };
+        case "arrayObjectString[2]":
+          return { str: "valid string" };
+        case "choice":
+          return "A";
+      }
+    };
+
+    const validation = await validate({
+      fields: [Schema.choice],
+      getValue,
+    });
+
+    expect(validation).toEqual([
+      { field: Schema.choice, error: "INVALID_VALUE" },
+      { field: Schema.arrayObjectString.nth(0), error: null },
+      { field: Schema.arrayObjectString.nth(1), error: "REQUIRED" },
+      { field: Schema.arrayObjectString.nth(2), error: null },
+    ]);
+  });
+
+  it("dependency change should not duplicate validation", async () => {
+    const required = (x: any) => (x ? null : "REQUIRED");
+
+    const { validate } = createFormValidator(Schema, validate => [
+      validate({
+        field: Schema.string,
+        rules: _number => [required],
+        dependencies: [Schema.number],
+      }),
+      validate({
+        field: Schema.number,
+        rules: () => [required],
+      }),
+    ]);
+
+    const getValue = () => "" as any;
+
+    const validation = await validate({
+      fields: [Schema.number, Schema.string],
+      getValue,
+    });
+
+    expect(validation).toEqual([
+      { field: Schema.number, error: "REQUIRED" },
+      { field: Schema.string, error: "REQUIRED" },
+    ]);
+  });
+
+  it("dependency change should trigger validation run even if root field triggers don't match", async () => {
+    const required = (x: any) => (x ? null : "REQUIRED");
+
+    const { validate } = createFormValidator(Schema, validate => [
+      validate({
+        field: Schema.string,
+        rules: _number => [required],
+        dependencies: [Schema.number],
+      }),
+      validate({
+        field: Schema.choice,
+        rules: _number => [required],
+        dependencies: [Schema.number],
+        triggers: ["blur"],
+      }),
+      validate({
+        field: Schema.number,
+        rules: () => [required],
+        triggers: ["blur"],
+      }),
+    ]);
+
+    const getValue = () => "" as any;
+
+    const validation = await validate({
+      fields: [Schema.number],
+      getValue,
+      trigger: "change",
+    });
+
+    expect(validation).toEqual([{ field: Schema.string, error: "REQUIRED" }]);
+  });
+
+  it("dependencies values are passed to rules constructor", async () => {
+    const { validate } = createFormValidator(Schema, validate => [
+      validate({
+        field: Schema.string,
+        rules: (...dependencies) => [
```

Comment:
this works but is somewhat convoluted

would be nice to create this function using `jest.fn` to then be able to make direct assertion that it was invoked with the dependencies